src/interface/state_descriptor.hpp

```diff
@@ -223,10 +200,51 @@ class StateDescriptor {
     return false;
   }
 
+
+  void PreFillDerived(std::shared_ptr<MeshBlockData<Real>> &rc) {
+    if (PreFillDerivedBlock != NULL) PreFillDerivedBlock(rc);
+  }
+  void PreFillDerived(std::shared_ptr<MeshData<Real>> &rc) {
+    if (PreFillDerivedMesh != NULL) PreFillDerivedMesh(rc);
+  }
+  void PostFillDerived(std::shared_ptr<MeshBlockData<Real>> &rc) {
+    if (PostFillDerivedBlock != NULL) PostFillDerivedBlock(rc);
+  }
+  void PostFillDerived(std::shared_ptr<MeshData<Real>> &rc) {
+    if (PostFillDerivedMesh != NULL) PostFillDerivedMesh(rc);
+  }
+  void FillDerived(std::shared_ptr<MeshBlockData<Real>> &rc) {
+    if (FillDerivedBlock != NULL) FillDerivedBlock(rc);
+  }
+  void FillDerived(std::shared_ptr<MeshData<Real>> &rc) {
+    if (FillDerivedMesh != NULL) FillDerivedMesh(rc);
+  }
+
+  Real EstimateTimestep(std::shared_ptr<MeshBlockData<Real>> &rc) {
+    if (EstimateTimestepBlock != NULL) return EstimateTimestepBlock(rc);
+    return std::numeric_limits<Real>::max();
+  }
+  Real EstimateTimestep(std::shared_ptr<MeshData<Real>> &rc) {
+    if (EstimateTimestepMesh != NULL) return EstimateTimestepMesh(rc);
+    return std::numeric_limits<Real>::max();
+  }
+
+  AmrTag CheckRefinement(std::shared_ptr<MeshBlockData<Real>> &rc) {
+    if (CheckRefinementBlock != NULL) return CheckRefinementBlock(rc);
+    return AmrTag::derefine;
+  }
```

Comment:
This is an elegant solution to the problem you were having before, and I like it a lot.